my-app-v/src/hooks/useTodoLogic.jsx

```diff
@@ -0,0 +1,198 @@
+        title = "Alert",
+        text = "Something happened!"
+    ) => {
+        Swal.fire({
+            icon: icon,
+            title: title,
+            text: text,
+            confirmButtonText: "OK"
+        });
+    };
+
+    // Fetch Todos
+    const { data: todos = [], isLoading, error } = useQuery({
+        queryKey: ["todos"],
+        queryFn: fetchTodos,
+    });
+
+    const addTodoMutation = useMutation({
+        mutationFn: addTodo,
+        onSuccess: () => {
+            queryClient.invalidateQueries(["todos"]);
+        },
+    });
+
+    const updateTodoMutation = useMutation({
+        mutationFn: updateTodo,
+        onSuccess: () => {
+            queryClient.invalidateQueries(["todos"]);
+        },
+    });
+
+    const deleteTodoMutation = useMutation({
+        mutationFn: deleteTodo,
+        onSuccess: () => {
+            queryClient.invalidateQueries(["todos"]);
+        },
+    });
+
+    const deleteTodoFun = useCallback((todo) => {
+        // not block delete if that todo item is not being editing.
+        if (editingTodo && editingTodo.id === todo.id) {
+            // make a alert to user that you are editing this todo item
+            //alert("You are editing this todo item. Please cancel editing to delete this item.");
+            fireAlertBySwel("warning", "Editing in Progress!", "You are editing this todo item. Please cancel editing to delete this item.");
+
+            return;
+        }
+
+        // asked user to confirm before deleting
+        Swal.fire({
+            title: "Do you want to delete the todo?",
+            showDenyButton: true,
+            confirmButtonText: "Delete",
+            denyButtonText: `cancel`
+        }).then((result) => {
+            if (result.isConfirmed) {
+                if (viewTodoItem.id === todo.id) {
+                    closeViewTodoComponent();
+                }
+                deleteTodoMutation.mutate({
+                    id: todo.id,
+                    key: todo.key,
+                    displayName: todo.displayName,
+                    description: todo.description,
+                    attribute: todo.attribute,
+                    completed: false,
+                });
+                Swal.fire("Delete!", "", "success");
+            }
+        });
+    }, [deleteTodoMutation]);
+
+    const handleChange = useCallback((e) => {
+        const { id, value } = e.target;
+        setFormData(prev => ({
+            ...prev,
+            [id]: value,
+        }));
+    }, []);
+
+    const validateTodoBeforeAdd = useCallback(() => {
+        const { key, displayName, description, attribute } = formData;
+
+        if (!key.trim() || !displayName.trim() || !description.trim() || !attribute.trim()) {
+            return { isValid: false, message: "All fields must be filled with non-empty values." };
+        }
+
+        if (editingTodo) {
+            return { isValid: true, message: "Validation successful." };
```

Comment:
[cc] what does `isValid` doing? we can simply return error message, if return is empty that means form is valid